app/assets/javascripts/discourse/app/models/private-message-topic-tracking-state.js

```diff
@@ -0,0 +1,174 @@
+import EmberObject from "@ember/object";
+import {
+  ARCHIVE_FILTER,
+  INBOX_FILTER,
+  NEW_FILTER,
+  UNREAD_FILTER,
+} from "discourse/routes/build-private-messages-route";
+import { NotificationLevels } from "discourse/lib/notification-levels";
+
+// See private_message_topic_tracking_state.rb for documentation
+const PrivateMessageTopicTrackingState = EmberObject.extend({
+  CHANNEL_PREFIX: "/private-message-topic-tracking-state",
+
+  newIncoming: Ember.A(),
+  inbox: null,
+  filter: null,
+  activeGroup: null,
+
+  init(opts) {
+    this._super(...arguments);
+    this.states = new Map();
+    this._loadStates(opts.data);
+    this.establishChannels();
+  },
+
+  establishChannels() {
+    this.messageBus.subscribe(
+      `${this.CHANNEL_PREFIX}/${this.user.id}`,
+      this._processMessage.bind(this)
+    );
+  },
+
+  lookupCount(type) {
+    const typeFilterFn = type === "new" ? this._isNew : this._isUnread;
+    let filterFn;
+
+    if (this.inbox === "user") {
+      filterFn = this._isPersonal.bind(this);
+    } else if (this.inbox === "group") {
+      filterFn = this._isGroup.bind(this);
+    }
+
+    return Array.from(this.states.values()).filter((topic) => {
+      return typeFilterFn(topic) && (!filterFn || filterFn(topic));
+    }).length;
+  },
+
+  trackIncoming(inbox, filter, group) {
+    this.setProperties({ inbox, filter, activeGroup: group });
+  },
+
+  resetTracking() {
+    if (this.inbox) {
+      this.set("newIncoming", Ember.A());
+    }
+  },
+
+  _isNew(topic) {
+    return (
+      topic.last_read_post_number === null &&
```

Comment:
do we need explicit check on null here ?

```
!topic.last_read_post_number && ...
```

Reply:
I updated to what @jjaffeux suggested because that reads much better to me.